File: others/400-500/410. 分割数组的最大值.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *

MOD = int(1e9 + 7)
DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
INF = int(1e9)


class Solution:
    def splitArray(self, nums: List[int], K: int) -> int:
        n = len(nums)

        @cache
        def dfs(i: int, k: int) -> int:
            if k == 1:
                return sum(nums[i:])
            res = INF
            s = 0
            for j in range(i, n - k + 1):
                s += nums[j]
                res = min(res, max(s, dfs(j + 1, k - 1)))
            return res

        return dfs(0, K)


class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        n = len(nums)
        f = [[10**18] * (m + 1) for _ in range(n + 1)]
        sub = [0]
        for elem in nums:
            sub.append(sub[-1] + elem)

        f[0][0] = 0
        for i in range(1, n + 1):
            for j in range(1, min(i, m) + 1):
                for k in range(i):
                    f[i][j] = min(f[i][j], max(f[k][j - 1], sub[i] - sub[k]))

        return f[n][m]


class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        def check(x: int) -> bool:
            total, cnt = 0, 1
            for num in nums:
                if total + num > x:
                    cnt += 1
                    total = num
                else:
                    total += num
            return cnt <= m

        left = max(nums)
        right = sum(nums)
        while left < right:
            mid = (left + right) // 2
            if check(mid):
                right = mid
            else:
                left = mid + 1

        return left

File: leetcode/400-500/410. 分割数组的最大值.cpp
//
// Created by 123 on 25-5-30.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

class Solution {
public:
	int splitArray(vector<int>& nums, int k) {
		int n = nums.size();
		int right = reduce(nums.begin(), nums.end(), 0) + 1;
		int left = ranges::max(nums);

		auto check = [&](int limit) -> bool {
			int count = 1;
			int i = 0;
			int sum = 0;
			while (i < n) {
				int x = nums[i];
				if (sum + x > limit) {
					sum = x;
					i++;
					count++;
				} else {
					sum += x;
					i++;
				}
				if (count > k) {
					return false;
				}
			}
			return true;
		};

		int ans = -1;
		while (left < right) {
			int mid = left + (right - left) / 2;
			if (check(mid)) {
				ans = mid;
				right = mid;
			} else {
				left = mid + 1;
			}
		}
		return ans;
	}
};

class Solution {
public:
	int splitArray(vector<int> &nums, int k) {
		int left = *max_element(nums.begin(), nums.end());
		int right = accumulate(nums.begin(), nums.end(), 0);

		auto check = [&](int mx) -> bool {
			int cnt = 1, sum = 0;
			for (int x: nums) {
				if (sum + x <= mx) {
					sum += x;
				} else {
					cnt++;
					sum = x;
				}
			}
			return cnt <= k;
		};

		int ans = right;
		while (left < right) {
			int mid = left + (right - left) / 2;
			if (check(mid)) {
				ans = mid;
				right = mid; // 尝试更小
			} else {
				left = mid + 1; // 说明 mid 太小
			}
		}
		return ans;
	}
};
